Provide initial functionality

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR implements the initial functionality for this repository, namely Terraform to do the following:
* Deploy a Lambda function, specifically [`cisagov/publish-egress-ip-lambda`](https://github.com/cisagov/publish-egress-ip-lambda), and it's execution role/policy
* Create the CloudWatch log group used by the Lambda
* Schedule the Lambda to run every X minutes using a CloudWatch event
* Create an S3 bucket to host the published files (objects) containing egress IP information
* Deploy a CloudFront distribution to enable HTTPS access to the egress IP info in the S3 bucket
* Create DNS records that point to the CloudFront distribution

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We need a mechanism to publish egress IP info from a subset of our AWS accounts and this Terraform will allow us to deploy the infrastructure to do that.

This is part of the work for:
* https://github.com/cisagov/cool-system/issues/208
* https://github.com/cisagov/cyhy_amis/issues/284

This PR is related to the following:
* https://github.com/cisagov/cool-accounts/pull/126
* https://github.com/cisagov/cool-accounts/pull/127

This PR resolves #5 (via https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/f22ab6d85407aeb82b0eaa5bb6d0baecd9f51f35).

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I successfully applied all of this Terraform and confirmed that the Lambda was deployed, scheduled, and worked as expected.  I also confirmed that I could cleanly destroy all of this Terraform.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
- [x] Tear down and redeploy all Terraform to ensure it all applies cleanly.
- [x] Ensure that https://github.com/cisagov/publish-egress-ip-lambda/pull/4 has been approved and merged.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Confirm that all internal stakeholders are ready for this deployment
- [x] Deploy to Production

dav3r committed:
Remove unneeded files inherited from skeleton

dav3r committed:
Define initial set of variables and remove unneded variables from skeleton

dav3r committed:
Define providers

dav3r committed:
Define Terraform backend

dav3r committed:
Define Terraform remote states

dav3r committed:
Define useful locals

dav3r committed:
Create S3 bucket where published file(s) will be stored

dav3r committed:
Define Terraform output and remove skeleton outputs

dav3r committed:
Add IAM policy and role used by the publish-egress-ip Lambda

dav3r committed:
Define the Lambda function to publish egress IP addresses

The ZIP file is created with http://github.com/cisagov/publish-egress-ip-lambda

Also:
* Define the CloudWatch log group used by the Lambda
* Schedule the Lambda function to run every X minutes using a CloudWatch event (configured via var.lambda_schedule_interval)

dav3r committed:
Set up CloudFront distribution and related S3/Lambda resources

These allow us to use an HTTPS endpoint, which S3 websites do not support natively.

dav3r committed:
Define DNS records that support CloudFront endpoints

dav3r committed:
Replace skeleton README with useful documentation

Also, add angle brackets around a bare URL that was causing the markdownlint pre-commit hook to fail.

dav3r committed:
Add gitignore entry for default lambda deployment package name

dav3r committed:
Add link to relevant issue for TODO comment

dav3r committed:
Add gitignore entry for "add security headers" lambda zip file

This file is automatically generated by Terraform, so we add it to .gitignore in order to minimize confusion.

jsf9k commented on route53.tf:
```diff
@@ -0,0 +1,29 @@
+  alias {
+    evaluate_target_health = false
+    name                   = aws_cloudfront_distribution.rules_s3_distribution.domain_name
+    zone_id                = "Z2FDTNDATAQYW2"
```
Is there any way to avoid hard-coding these Route53 zone IDs?

dav3r replied:
Not that I know of, but please take a look at https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/5f5135fa2615c1456463493b76eaaf5fa5e882c5 and let me know if you think that helps.

jsf9k replied:
Ah, that is much better.  Thank you.

jsf9k reviewed (commented):
Nice work!  I have one question before approving.

dav3r committed:
Clarify the standard CloudFront Route53 zone ID

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
Powerful work!

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| deployment\_role\_arn | The ARN of the IAM role to use to deploy the Lambda and all related resources. | `string` | n/a | yes |
+| domain | The domain hosting the published file(s) containing egress IPs.  Also used for the CloudFront distribution and certificate. | `string` | n/a | yes |
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query. | `string` | `"EC2ReadOnly"` | no |
+| extraorg\_account\_ids | A list of AWS account IDs corresponding to "extra" accounts that you want to query for egress IPs to publish. | `list(string)` | `[]` | no |
```
Is it ok if these account IDs include accounts that match the regex pattern supplied in `account_name_regex`?

dav3r replied:
That's fine since I'm creating a set with all of the regex-matching accounts and the "extraorg" accounts, so any duplicated account IDs won't be an issue:

https://github.com/cisagov/publish-egress-ip-terraform/blob/5f5135fa2615c1456463493b76eaaf5fa5e882c5/locals.tf#L27-L31

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| domain | The domain hosting the published file(s) containing egress IPs.  Also used for the CloudFront distribution and certificate. | `string` | n/a | yes |
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query. | `string` | `"EC2ReadOnly"` | no |
+| extraorg\_account\_ids | A list of AWS account IDs corresponding to "extra" accounts that you want to query for egress IPs to publish. | `list(string)` | `[]` | no |
+| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An examp ...
```
```suggestion
| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An example file configuration looks like this: `[{"app\_regex": ".*", "description": "This file contains a list of all public IP addresses to be published.", "filename": "all.txt",  "static\_ips": []}, {"app\_regex": "^Vulnerability Scanning$", "description": "This file contains a list of all IPs used for Vulnerability Scanning.", "filename": "vs.txt",  "static\_ips": ["192.168.1.1/32", "192.168.2.2/32"]}]` | `list(object({ app_regex = string, description = string, filename = string, static_ips = list(string) }))` | `[]` | no |
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query. | `string` | `"EC2ReadOnly"` | no |
+| extraorg\_account\_ids | A list of AWS account IDs corresponding to "extra" accounts that you want to query for egress IPs to publish. | `list(string)` | `[]` | no |
+| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An examp ...
+| file\_header | The header template for each published file.  The following variables are available within the template: {domain} - the domain where the published files are located, {filename} - the name of the published file, {timestamp} - the timestamp when the file was published, {description} - the description of the published file | `string` | `"###\n# https://{domain}/{filename}n# {timesta ...
```
I'm not sure if this is a missed update or a problem with the auto-generated table. The definition in `variables.tf` includes the missing backslash.

```suggestion
| file\_header | The header template for each published file.  The following variables are available within the template: {domain} - the domain where the published files are located, {filename} - the name of the published file, {timestamp} - the timestamp when the file was published, {description} - the description of the published file | `string` | `"###\n# https://{domain}/{filename}\n# {timestamp}\n# {description}\n###\n"` | no |
```

dav3r replied:
BTW @mcdonnnj, I just double-checked and this appears to be a bug in the table generation by `terraform-docs`.  Something to keep an eye out for in the future.

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| lambda\_function\_description | The description of the Lambda function. | `string` | `"Lambda function to publish egress IP addresses to an S3 bucket configured with a CloudFront distribution for HTTPS access."` | no |
+| lambda\_function\_name | The name of the Lambda function to publish egress IP addresses. | `string` | `"publish-egress-ip"` | no |
+| lambda\_schedule\_interval | The number of minutes betweeen scheduled runs of the Lambda function to publish egress IP addresses. | `string` | `"60"` | no |
+| lambda\_zip\_filename | The name of the ZIP file containing the Lambda function to publish egress IP addresses.  The file must be located in the root directory of this project. | `string` | `"publish_egress_ip.zip"` | no |
```
```suggestion
| lambda\_zip\_filename | The name of the ZIP file containing the Lambda function deployment package to publish egress IP addresses.  The file must be located in the root directory of this project. | `string` | `"publish_egress_ip.zip"` | no |
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| lambdaexecution\_role\_description | The description to associate with the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"Allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and ...
+| lambdaexecution\_role\_name | The name to assign the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"PublishEgressIPLambda"` | no |
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: [{ "Name" : "endpoint", "Values" : ["*.us-*"] }].  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>  ...
```
```suggestion
| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html> | `list(object({ Name = string, Values = list(string) }))` | `[]` | no |
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: [{ "Name" : "endpoint", "Values" : ["*.us-*"] }].  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>  ...
+| root\_object | The root object in the S3 bucket to serve when no path is provided or an error occurs. | `string` | `"all.txt"` | no |
+| route53\_role\_arn | The ARN of the IAM role to use to modify route53 DNS resources. | `string` | n/a | yes |
```
```suggestion
| route53\_role\_arn | The ARN of the IAM role to use to modify Route53 DNS resources. | `string` | n/a | yes |
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: [{ "Name" : "endpoint", "Values" : ["*.us-*"] }].  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>  ...
+| root\_object | The root object in the S3 bucket to serve when no path is provided or an error occurs. | `string` | `"all.txt"` | no |
+| route53\_role\_arn | The ARN of the IAM role to use to modify route53 DNS resources. | `string` | n/a | yes |
+| tags | Tags to apply to all AWS resources created | `map(string)` | `{}` | no |
```
```suggestion
| tags | Tags to apply to all AWS resources created. | `map(string)` | `{}` | no |
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+variable "file_configs" {
+  default     = []
+  type        = list(object({ app_regex = string, description = string, filename = string, static_ips = list(string) }))
+  description = "A list of objects that define the files to be published.  \"app_regex\" specifies a regular expression matching the application name (based on the var.application_tag).  \"description\" is the description of the published file.  \"filename\" is the name to assign the published file.  \"static_ips\" is a list of CIDR blocks that will always be included in the published file.  An e ...
```
```suggestion
  description = "A list of objects that define the files to be published.  \"app_regex\" specifies a regular expression matching the application name (based on the var.application_tag).  \"description\" is the description of the published file.  \"filename\" is the name to assign the published file.  \"static_ips\" is a list of CIDR blocks that will always be included in the published file.  An example file configuration looks like this: `[{\"app_regex\": \".*\", \"description\": \"This file contains a list of all public IP addresses to be published.\", \"filename\": \"all.txt\",  \"static_ips\": []}, {\"app_regex\": \"^Vulnerability Scanning$\", \"description\": \"This file contains a list of all IPs used for Vulnerability Scanning.\", \"filename\": \"vs.txt\",  \"static_ips\": [\"192.168.1.1/32\", \"192.168.2.2/32\"]}]`"
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+
+variable "lambda_zip_filename" {
+  default     = "publish_egress_ip.zip"
+  description = "The name of the ZIP file containing the Lambda function to publish egress IP addresses.  The file must be located in the root directory of this project."
```
```suggestion
  description = "The name of the ZIP file containing the Lambda function deployment package to publish egress IP addresses.  The file must be located in the root directory of this project."
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+
+variable "region_filters" {
+  default     = []
+  description = "A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: [{ \"Name\" : \"endpoint\", \"Values\" : [\"*.us-*\"] }].  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.h ...
```
```suggestion
  description = "A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ \"Name\" : \"endpoint\", \"Values\" : [\"*.us-*\"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>"
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+
+variable "tags" {
+  type        = map(string)
+  description = "Tags to apply to all AWS resources created"
```
```suggestion
  description = "Tags to apply to all AWS resources created."
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+}
+
+variable "route53_role_arn" {
+  description = "The ARN of the IAM role to use to modify route53 DNS resources."
```
```suggestion
  description = "The ARN of the IAM role to use to modify Route53 DNS resources."
```

mcdonnnj commented on cloudfront.tf:
```diff
@@ -0,0 +1,152 @@
+  source  = "transcend-io/lambda-at-edge/aws"
+  version = "0.5.0"
+
+  description            = "Adds HSTS and other security headers to the response"
```
```suggestion
  description            = "Adds HSTS and other security headers to the response."
```

mcdonnnj commented on cloudfront.tf:
```diff
@@ -0,0 +1,152 @@
+
+  restrictions {
+    geo_restriction {
+      locations        = ["US", "PR", "VI", "AS", "GU", "MP"]
```
I'm assuming order does not matter for functionality so I would prefer alphabetical sorting.

```suggestion
      locations        = ["AS", "GU", "MP", "PR", "US", "VI"]
```

mcdonnnj commented on lambda.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_lambda_function" "publish_egress_ip" {
+  provider = aws.deploy
+
+  filename         = var.lambda_zip_filename
+  source_code_hash = filebase64sha256(var.lambda_zip_filename)
+  function_name    = var.lambda_function_name
+  role             = aws_iam_role.lambdaexecution_role.arn
+  handler          = "lambda_handler.handler"
+  runtime          = "python3.9"
+  timeout          = 600
+  memory_size      = 128
+  description      = var.lambda_function_description
```
Just sorting.

```suggestion
  description      = var.lambda_function_description
  filename         = var.lambda_zip_filename
  function_name    = var.lambda_function_name
  handler          = "lambda_handler.handler"
  memory_size      = 128
  role             = aws_iam_role.lambdaexecution_role.arn
  runtime          = "python3.9"
  source_code_hash = filebase64sha256(var.lambda_zip_filename)
  timeout          = 600
```

mcdonnnj commented on lambda_assume_role_policy_doc.tf:
```diff
@@ -0,0 +1,17 @@
+data "aws_iam_policy_document" "lambda_assume_role_doc" {
+  statement {
+    actions = [
+      "sts:AssumeRole",
```
```suggestion
      "sts:AssumeRole",
      "sts:TagSession",
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+}
+
+variable "lambda_schedule_interval" {
+  default     = "60"
```
```suggestion
  default     = 60
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,130 @@
+variable "lambda_schedule_interval" {
+  default     = "60"
+  description = "The number of minutes betweeen scheduled runs of the Lambda function to publish egress IP addresses."
+  type        = string
```
```suggestion
  type        = number
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| file\_header | The header template for each published file.  The following variables are available within the template: {domain} - the domain where the published files are located, {filename} - the name of the published file, {timestamp} - the timestamp when the file was published, {description} - the description of the published file | `string` | `"###\n# https://{domain}/{filename}n# {timesta ...
+| lambda\_function\_description | The description of the Lambda function. | `string` | `"Lambda function to publish egress IP addresses to an S3 bucket configured with a CloudFront distribution for HTTPS access."` | no |
+| lambda\_function\_name | The name of the Lambda function to publish egress IP addresses. | `string` | `"publish-egress-ip"` | no |
+| lambda\_schedule\_interval | The number of minutes betweeen scheduled runs of the Lambda function to publish egress IP addresses. | `string` | `"60"` | no |
```
Corresponding suggestions are in `variables.tf` but I think this value should be a `number`. I verified that the existing use in `format()` calls work as written in the Terraform console.

```suggestion
| lambda\_schedule\_interval | The number of minutes betweeen scheduled runs of the Lambda function to publish egress IP addresses. | `number` | `60` | no |
```

mcdonnnj reviewed (commented):
Some suggestions/thoughts on my first pass. I *think* I have suggestions in `variables.tf` corresponding to all suggestions in `README.md`.

dav3r commented on lambda.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_cloudwatch_event_rule" "lambda_schedule" {
+  provider = aws.deploy
+
+  description         = format("Executes %s Lambda every %s minutes", var.lambda_function_name, var.lambda_schedule_interval)
+  name                = format("%s-every-%s-minutes", var.lambda_function_name, var.lambda_schedule_interval)
+  schedule_expression = format("rate(%s minutes)", var.lambda_schedule_interval)
```
Use `%d` instead of `%s` to correspond to `var.lambda_schedule_interval` becoming a number:
```suggestion
  description         = format("Executes %s Lambda every %d minutes", var.lambda_function_name, var.lambda_schedule_interval)
  name                = format("%s-every-%d-minutes", var.lambda_function_name, var.lambda_schedule_interval)
  schedule_expression = format("rate(%d minutes)", var.lambda_schedule_interval)
```

mcdonnnj replied:
So a note here is that this controls what the value is converted to per the [documentation](https://www.terraform.io/language/functions/format#verbs) and does not specify the provided type. We should use `%d` anywhere we expect/need an integer value (or whatever the appropriate `%` verb is for what we expect/need) no matter what type the source is to ensure behavior. Examples from the console:

```console
> format("%s %d", "60.15", "60")
"60.15 60"
> format("%s %d", 60.15, 60.15)
╷
│ Error: Error in function call
│
│   on <console-input> line 1:
│   (source code not available)
│
│ Call to function "format" failed: unsupported value for "%d" at 3: an integer is required.
╵


> format("%s %d", 60, 60)
"60 60"
> format("%s %d", 60, "60")
"60 60"
```

dav3r replied:
It's a little bit ugly (if you have a better solution, please let me know), but take a look at the validation I added in https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/b4755aeeafc683a5cf6abb0294aa92b7d1609446.

dav3r committed:
Make lambda schedule interval a number instead of a string

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Correct capitalization and grammar

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use more accurate verbiage in a variable description

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Sort some thangs alphabetically

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add some helpful backticks to variable descriptions

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add a missing backslash character

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Make default lambda schedule interval a number

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add a missing period

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add TagSession permission to Lambda assume role

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Correct a typo in a variable description

dav3r committed:
Add validation to lambda_schedule_interval variable

The validation ensures that a positive integer has been provided.

mcdonnnj commented on lambda.tf:
```diff
@@ -0,0 +1,68 @@
+
+  description         = format("Executes %s Lambda every %d minutes", var.lambda_function_name, var.lambda_schedule_interval)
+  name                = format("%s-every-%d-minutes", var.lambda_function_name, var.lambda_schedule_interval)
+  schedule_expression = format("rate(%d minutes)", var.lambda_schedule_interval)
```
Per the documentation:
> If the value is equal to 1, then the unit must be singular. Similarly, for values greater than 1, the unit must be plural. For example, rate(1 hours) and rate(5 hour) are not valid, but rate(1 hour) and rate(5 hours) are valid.

If you have another way you'd prefer to do this feel free.

```suggestion
  schedule_expression = format("rate(%d minute%s)", var.lambda_schedule_interval, var.lambda_schedule_interval != 1 ? "s" : "")
```

dav3r replied:
That works for me- nice catch!  After seeing your suggestion, I briefly thought about restricting `var.lambda_schedule_interval` to be greater than 1, but that's pretty lame. 😄

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,135 @@
+  type        = number
+
+  validation {
+    condition     = try(parseint(tostring(var.lambda_schedule_interval), 10), null) == var.lambda_schedule_interval && var.lambda_schedule_interval > 0
```
Since the value has already been converted to a numeric value I think this makes more sense than a series of type conversions. Alternatively I think `can(parseint(tostring(var.lambda_schedule_interval), 10))` is a better implementation for the first test. I subbed in `alltrue()` because I think it makes for a cleaner representation of a series of conditions.

```suggestion
    condition     = alltrue([floor(var.lambda_schedule_interval) == var.lambda_schedule_interval, var.lambda_schedule_interval > 0])
```

dav3r replied:
I think your suggestion is absolutely an improvement over what I had initially, thanks!

BTW, I didn't know that `can()` function existed- that is what I was looking for earlier when I used `try()`.  Good to know!

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| lambda\_zip\_filename | The name of the ZIP file containing the Lambda function deployment package to publish egress IP addresses.  The file must be located in the root directory of this project. | `string` | `"publish_egress_ip.zip"` | no |
+| lambdaexecution\_role\_description | The description to associate with the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"Allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and ...
+| lambdaexecution\_role\_name | The name to assign the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"PublishEgressIPLambda"` | no |
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
```
Nitpicky but I prefer to reference that the tag is on a resource.

```suggestion
| publish\_egress\_tag | The name of the AWS resource tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,135 @@
+
+variable "publish_egress_tag" {
+  default     = "Publish Egress"
+  description = "The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published."
```
Mirroring the suggestion in `README.md`.

```suggestion
  description = "The name of the AWS resource tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published."
```

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| lambdaexecution\_role\_description | The description to associate with the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"Allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and ...
+| lambdaexecution\_role\_name | The name to assign the IAM role (and policy) that allows the publish-egress-ip Lambda to query other accounts for public EC2 IP information, publish objects to the S3 bucket, and write CloudWatch logs. | `string` | `"PublishEgressIPLambda"` | no |
+| publish\_egress\_tag | The name of the AWS tag whose value represents whether the EC2 instance or elastic IP should have its public IP address published. | `string` | `"Publish Egress"` | no |
+| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html ...
```
```suggestion
| region\_filters | A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ "Name" : "endpoint", "Values" : ["*.us-*"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>. | `list(object({ Name = string, Values = list(string) }))` | `[]` | no |
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,135 @@
+
+variable "region_filters" {
+  default     = []
+  description = "A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ \"Name\" : \"endpoint\", \"Values\" : [\"*.us-*\"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions ...
```
```suggestion
  description = "A list of AWS EC2 region filters to use when querying for IP addresses to publish.  If a filter is not specified, the query will be performed in all regions.  An example filter to restrict to US regions looks like this: `[{ \"Name\" : \"endpoint\", \"Values\" : [\"*.us-*\"] }]`.  For more information, refer to <https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-regions.html>."
```

mcdonnnj commented on cloudfront.tf:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  aliases             = [var.domain]
+  comment             = "Created by cisagov/publish-egress-ip-terraform"
```
I like periods a lot.

```suggestion
  comment             = "Created by cisagov/publish-egress-ip-terraform."
```

mcdonnnj commented on lambda.tf:
```diff
@@ -0,0 +1,68 @@
+  retention_in_days = 30
+}
+
+# Schedule the Lambda function to run every X minutes
```
```suggestion
# Schedule the Lambda function to run every X minute(s)
```

mcdonnnj commented on lambda.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_cloudwatch_event_rule" "lambda_schedule" {
+  provider = aws.deploy
+
+  description         = format("Executes %s Lambda every %d minutes", var.lambda_function_name, var.lambda_schedule_interval)
```
```suggestion
  description         = format("Executes %s Lambda every %d minute(s).", var.lambda_function_name, var.lambda_schedule_interval)
```

mcdonnnj commented on providers.tf:
```diff
@@ -1,13 +1,50 @@
+}
+
+# The provider used to create resources inside the AWS account where
+# the lambda and S3 bucket will be deployed.
```
```suggestion
# the Lambda and S3 bucket will be deployed.
```

mcdonnnj commented on providers.tf:
```diff
@@ -1,13 +1,50 @@
+  region = var.aws_region
+}
+
+# The provider that can modify route53 resources in the DNS account.
```
```suggestion
# The provider that can modify Route53 resources in the DNS account.
```

mcdonnnj reviewed (approved):
Thanks for your continued work on this. I have for your consideration some suggestions around the validation you added for the `lambda_schedule_interval` variable, correctly defining the rate provided to the `schedule_expression` argument, and then some miscellaneous capitalization/language suggestions.

dav3r committed:
Improve language, grammar, and capitalization

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Ensure Lambda schedule_expression uses correct syntax in all cases

According to the documentation, "If the value is equal to 1, then the unit must be singular."

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve validation test for lambda_schedule_interval variable

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
I just added the `blocked` label on this PR so that I don't merge it until after https://github.com/cisagov/publish-egress-ip-lambda/pull/4 has been approved and merged.  I also added a `Pre-merge checklist` checkbox to the PR description above.

mcdonnnj commented on README.md:
```diff
@@ -42,51 +82,88 @@ module "example" {
+| domain | The domain hosting the published file(s) containing egress IPs.  Also used for the CloudFront distribution and certificate. | `string` | n/a | yes |
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query. | `string` | `"EC2ReadOnly"` | no |
+| extraorg\_account\_ids | A list of AWS account IDs corresponding to "extra" accounts that you want to query for egress IPs to publish. | `list(string)` | `[]` | no |
+| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An examp ...
```
I missed this when suggesting this be quoted but the underscores don't need to be escaped.

```suggestion
| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An example file configuration looks like this: `[{"app_regex": ".*", "description": "This file contains a list of all public IP addresses to be published.", "filename": "all.txt",  "static_ips": []}, {"app_regex": "^Vulnerability Scanning$", "description": "This file contains a list of all IPs used for Vulnerability Scanning.", "filename": "vs.txt",  "static_ips": ["192.168.1.1/32", "192.168.2.2/32"]}]` | `list(object({ app_regex = string, description = string, filename = string, static_ips = list(string) }))` | `[]` | no |
```

dav3r committed:
Remove unnecessary escaping of underscores

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Make file_header a list of strings, instead of a string

When the file is published, newline characters are added between each item in the list.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (commented):
Quick reminder to enable these ignore directives:
https://github.com/cisagov/publish-egress-ip-terraform/blob/d1c0a2906d3996c536eac8d198d23efc7175c303/.github/dependabot.yml#L24-L36

dav3r committed:
Enable dependabot ignore directives

These are commented out in our skeletons, but we want to ignore these alerts in downstream repos like this one.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
> Quick reminder to enable these ignore directives

Thanks @mcdonnnj for the reminder!  Done in https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/3d83004aaf2e0796645bc64a8e4605548acefa1e.

dav3r committed:
Satisfy markdown linter

dav3r committed:
Alphabetize resource arguments

dav3r committed:
Update README with latest output from terraform-docs

dav3r committed:
Remove lifecycle blocks

These are no longer needed now that we have moved to version 4.x of the Terraform AWS provider.

dav3r committed:
Modify name of expected Lambda deployment file to match our current standard name

dav3r committed:
Remove a file that was inherited from the skeleton

dav3r committed:
Specify the runtime to use for the lambda-at-edge module

Using the current default runtime (nodejs14.x) of this module results in this error: "The runtime parameter of nodejs14.x is no longer supported for creating or updating AWS Lambda functions. We recommend you use the new runtime (nodejs20.x) while creating or updating functions."

Using the AWS-recommended runtime of nodejs20.x results in this error: "expected runtime to be one of [nodejs nodejs4.3 nodejs6.10 nodejs8.10 nodejs10.x nodejs12.x nodejs14.x nodejs16.x java8 java8.al2 java11 python2.7 python3.6 python3.7 python3.8 python3.9 dotnetcore1.0 dotnetcore2.0 dotnetcore2.1 dotnetcore3.1 dotnet6 nodejs4.3-edge go1.x ruby2.5 ruby2.7 provided provided.al2 nodejs18.x python3.10 java17], got nodejs20.x"

As best I can tell, this is because the Terraform AWS provider that we are currently pinned to (~> 4.9) does not support nodejs20.x.

Given all of the info above, I chose to set the runtime to nodejs18.x and everything is working.

dav3r committed:
Update the mechanism for defining public read-only access to the egress bucket

Things have changed a bit with how AWS does public bucket access since this code was initially created, so we must change with the times.

dav3r committed:
Increase lambda timeout from 10 to 15 minutes

10 minutes was not long enough to iterate through all of our AWS accounts.

jsf9k commented on variables.tf:
```diff
@@ -31,3 +47,92 @@ variable "aws_region" {
+variable "file_configs" {
+  default     = []
+  type        = list(object({ app_regex = string, description = string, filename = string, static_ips = list(string) }))
+  description = "A list of objects that define the files to be published.  \"app_regex\" specifies a regular expression matching the application name (based on the var.application_tag).  \"description\" is the description of the published file.  \"filename\" is the name to assign the published file.  \"static_ips\" is a list of CIDR blocks that will always be included in the published file.  An e ...
```
```suggestion
  description = "A list of objects that define the files to be published.  \"app_regex\" specifies a regular expression matching the application name (based on the variable var.application_tag).  \"description\" is the description of the published file.  \"filename\" is the name to assign the published file.  \"static_ips\" is a list of CIDR blocks that will always be included in the published file.  An example file configuration looks like this: `[{\"app_regex\": \".*\", \"description\": \"This file contains a list of all public IP addresses to be published.\", \"filename\": \"all.txt\",  \"static_ips\": []}, {\"app_regex\": \"^Vulnerability Scanning$\", \"description\": \"This file contains a list of all IPs used for Vulnerability Scanning.\", \"filename\": \"vs.txt\",  \"static_ips\": [\"192.168.1.1/32\", \"192.168.2.2/32\"]}]`"
```

jsf9k reviewed (approved):
I have one suggested language tweak, but nothing that should hold up this PR.

dav3r commented on README.md:
```diff
@@ -43,37 +83,81 @@ module "example" {
+| domain | The domain hosting the published file(s) containing egress IPs.  Also used for the CloudFront distribution and certificate. | `string` | n/a | yes |
+| ec2\_read\_role\_name | The name of the IAM role that allows read access to the necessary EC2 attributes.  Note that this role must exist in each account that you want to query. | `string` | `"EC2ReadOnly"` | no |
+| extraorg\_account\_ids | A list of AWS account IDs corresponding to "extra" accounts that you want to query for egress IPs to publish. | `list(string)` | `[]` | no |
+| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An examp ...
```
```suggestion
| file\_configs | A list of objects that define the files to be published.  "app\_regex" specifies a regular expression matching the application name (based on the variable var.application\_tag).  "description" is the description of the published file.  "filename" is the name to assign the published file.  "static\_ips" is a list of CIDR blocks that will always be included in the published file.  An example file configuration looks like this: `[{"app_regex": ".*", "description": "This file contains a list of all public IP addresses to be published.", "filename": "all.txt",  "static_ips": []}, {"app_regex": "^Vulnerability Scanning$", "description": "This file contains a list of all IPs used for Vulnerability Scanning.", "filename": "vs.txt",  "static_ips": ["192.168.1.1/32", "192.168.2.2/32"]}]` | `list(object({ app_regex = string, description = string, filename = string, static_ips = list(string) }))` | `[]` | no |
```

dav3r committed:
Improve a variable description

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj commented on s3.tf:
```diff
@@ -0,0 +1,90 @@
+resource "aws_s3_bucket_public_access_block" "egress_info" {
+  provider = aws.deploy
+
+  bucket = aws_s3_bucket.egress_info.id
+
+  block_public_acls       = false
+  block_public_policy     = false
+  ignore_public_acls      = false
+  restrict_public_buckets = false
```
This change is possible if functionality like https://github.com/cisagov/cyhy_amis/pull/742 is incorporated. I think it is better to have this bucket's contents private and let the CloudFront distribution exclusively handle serving the desired files.

```suggestion
  block_public_acls       = true
  block_public_policy     = true
  bucket                  = aws_s3_bucket.egress_info.id
  ignore_public_acls      = true
  restrict_public_buckets = true
```

dav3r replied:
Thanks for pointing this out- I applied your suggestion in https://github.com/cisagov/publish-egress-ip-terraform/pull/6/commits/bc3338e51d99eb54987b374558b3ebd14a4d297e.  Please let me know if you see anything in there that doesn't make sense.

dav3r committed:
Replace bucket "public-read" ACL with a policy that only allows CloudFront to read the bucket contents

This is a more secure bucket configuration, since there is no direct public access to the bucket.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on README.md:
````diff
@@ -2,33 +2,73 @@
+- A Terraform [variables](variables.tf) file customized for your
+  assessment environment, for example:
+
+  ```console
````
I believe this is still a valid formatting?

```suggestion
  ```hcl
```

mcdonnnj commented on lambda_policy.tf:
```diff
@@ -0,0 +1,55 @@
+  statement {
+    actions = [
+      "s3:PutObject",
+      "s3:PutObjectAcl",
+      "s3:PutObjectVersionAcl",
```
I don't think these are needed anymore.

```suggestion
```

mcdonnnj commented on route53.tf:
```diff
@@ -0,0 +1,29 @@
+resource "aws_route53_record" "rules_vm_A" {
+  provider = aws.route53resourcechange
+
+  alias {
+    evaluate_target_health = false
+    name                   = aws_cloudfront_distribution.egress_info.domain_name
+    zone_id                = local.cloudfront_zone_id
+  }
+  name    = var.domain
+  type    = "A"
+  zone_id = data.terraform_remote_state.dns_cyber_dhs_gov.outputs.cyber_dhs_gov_zone.id
```
Based on team consensus on Terraform resource forrmatting.

```suggestion
  name    = var.domain
  type    = "A"
  zone_id = data.terraform_remote_state.dns_cyber_dhs_gov.outputs.cyber_dhs_gov_zone.id

  alias {
    evaluate_target_health = false
    name                   = aws_cloudfront_distribution.egress_info.domain_name
    zone_id                = local.cloudfront_zone_id
  }
```

mcdonnnj commented on route53.tf:
```diff
@@ -0,0 +1,29 @@
+resource "aws_route53_record" "rules_vm_AAAA" {
+  provider = aws.route53resourcechange
+
+  alias {
+    evaluate_target_health = false
+    name                   = aws_cloudfront_distribution.egress_info.domain_name
+    zone_id                = local.cloudfront_zone_id
+  }
+  name    = var.domain
+  type    = "AAAA"
+  zone_id = data.terraform_remote_state.dns_cyber_dhs_gov.outputs.cyber_dhs_gov_zone.id
```
Based on team consensus on Terraform resource forrmatting.

```suggestion
  name    = var.domain
  type    = "AAAA"
  zone_id = data.terraform_remote_state.dns_cyber_dhs_gov.outputs.cyber_dhs_gov_zone.id

  alias {
    evaluate_target_health = false
    name                   = aws_cloudfront_distribution.egress_info.domain_name
    zone_id                = local.cloudfront_zone_id
  }
```

mcdonnnj reviewed (approved):
This looks pretty solid. Some minor items for your consideration. I didn't want to go wild with the formatting changes until we codify it in the development guide but left the two suggestions I'd already done.

dav3r committed:
Specify hcl formatting in tfvars example

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove now-unneeded ACL-related permissions

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Move TF configuration blocks to be after TF arguments in resources

This is a recently agreed upon change to how our team should format Terraform.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Alphabetically order TF arguments in a resource

dav3r committed:
Adjust some newlines for readability
